File: algorithms_py/inno_unionFind.py
class UnionFindV1HeavyFind:
    def __init__(self, size: int):
        self.root = [i for i in range(size)]

    def _find(self, x: int) -> int:
        return self.root[x]

    def union(self, x: int, y: int) -> bool:
        fx, fy = self._find(x), self._find(y)
        if fx == fy:
            return False
        for i in range(len(self.root)):
            if self.root[i] == fy:
                self.root[i] = fx
        return True

    def connected(self, x: int, y: int) -> bool:
        return self._find(x) == self._find(y)

    def componentCount(self) -> int:
        ans = 0
        for i, item in enumerate(self.root):
            if i == item:
                ans += 1
        return ans

    def printRoots(self):
        print(self.root)


class UnionFindV2HeavyUnion:
    def __init__(self, size: int):
        self.root = [i for i in range(size)]

    def _find(self, x: int) -> int:
        if self.root[x] == x:
            return x
        # flattens the tree
        self.root[x] = self._find(self.root[x])
        return self.root[x]

    def union(self, x: int, y: int) -> bool:
        fx, fy = self._find(x), self._find(y)
        if fx == fy:
            return False
        self.root[fy] = x
        return True

    def connected(self, x: int, y: int) -> bool:
        return self._find(x) == self._find(y)

    def componentCount(self) -> int:
        ans = 0
        for i, item in enumerate(self.root):
            if i == item:
                ans += 1
        return ans

    def printRoots(self):
        print(self.root)


class UnionFindV3ByRank:
    def __init__(self, size):
        self.root = [i for i in range(size)]
        self.rank = [1] * size

    def find(self, x) -> int:
        while x != self.root[x]:
            x = self.root[x]
        return x

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.root[rootX] = rootY
            else:
                self.root[rootY] = rootX
                self.rank[rootX] += 1

    def connected(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def componentCount(self) -> int:
        ans = 0
        for i, item in enumerate(self.root):
            if i == item:
                ans += 1
        return ans

    def printRoots(self):
        print(self.root)

File: algorithms_go/inno_unionfind.go
package algorithmsgo

import "fmt"

type InnoUnionFindHeavyFind struct {
	root []int
}

// Initializes a union find
func InitInnoUnionFindHeavyFind(size int) *InnoUnionFindHeavyFind {
	root := make([]int, size)
	for i := 0; i < size; i++ {
		root[i] = i
	}
	return &InnoUnionFindHeavyFind{root: root}
}

// Find the root node of x
func (uf *InnoUnionFindHeavyFind) Find(x int) int {
	return (*uf).root[x]
}

// Connect y into x
func (uf *InnoUnionFindHeavyFind) Union(x int, y int) bool {
	fx, fy := uf.Find(x), uf.Find(y)
	if fx == fy {
		return false
	}
	n := len(uf.root)
	for i := 0; i < n; i++ {
		if uf.root[i] == fy {
			uf.root[i] = fx
		}
	}
	return true
}

// Checks to see if x and y are connected
func (uf *InnoUnionFindHeavyFind) Connected(x int, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

// Get number of components
func (uf *InnoUnionFindHeavyFind) ComponentCount() int {
	ans := 0
	for i, item := range uf.root {
		if i == item {
			ans += 1
		}
	}
	return ans
}

// Prints the root
func (uf *InnoUnionFindHeavyFind) PrintRoot() {
	fmt.Println(uf.root)
}

type InnoUnionFindHeavyUnion struct {
	root []int
}

// Initializes a union find
func InitInnoUnionFindHeavyUnion(size int) *InnoUnionFindHeavyUnion {
	root := make([]int, size)
	for i := 0; i < size; i++ {
		root[i] = i
	}
	return &InnoUnionFindHeavyUnion{root: root}
}

// Find root
func (uf *InnoUnionFindHeavyUnion) Find(x int) int {
	if uf.root[x] == x {
		return x
	}
	uf.root[x] = uf.Find(uf.root[x])
	return uf.root[x]
}

// Union two components
func (uf *InnoUnionFindHeavyUnion) Union(x int, y int) bool {
	fx, fy := uf.Find(x), uf.Find(y)
	if fx == fy {
		return false
	}
	uf.root[fy] = x
	return true
}

// Checks if two nodes are connected
func (uf *InnoUnionFindHeavyUnion) Connected(x int, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

// Get component count
func (uf *InnoUnionFindHeavyUnion) ComponnetCount() int {
	ans := 0
	for i, item := range uf.root {
		if i == item {
			ans += 1
		}
	}
	return ans
}

// Print
func (uf *InnoUnionFindHeavyUnion) PrintRoots() {
	fmt.Println(uf.root)
}

type InnoUnionFindByRank struct {
	root []int
	rank []int
}

// Initializes a union find
func InitInnoUnionFindByRank(size int) *InnoUnionFindByRank {
	root := make([]int, size)
	rank := make([]int, size)
	for i := 0; i < size; i++ {
		root[i] = i
		rank[i] = 1
	}
	return &InnoUnionFindByRank{root: root, rank: rank}
}

// Find root
func (uf *InnoUnionFindByRank) Find(x int) int {
	for x != uf.root[x] {
		x = uf.root[x]
	}
	return x
}

// Union two components
func (uf *InnoUnionFindByRank) Union(x int, y int) bool {
	rootX, rootY := uf.Find(x), uf.Find(y)
	if rootX == rootY {
		return false
	}
	if uf.rank[rootX] > uf.rank[rootY] {
		uf.root[rootY] = rootX
	} else if uf.rank[rootX] < uf.rank[rootY] {
		uf.root[rootX] = rootY
	} else {
		uf.root[rootY] = rootX
		uf.rank[rootX]++
	}

	return true
}

// Checks if two nodes are connected
func (uf *InnoUnionFindByRank) Connected(x int, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

// Get component count
func (uf *InnoUnionFindByRank) ComponnetCount() int {
	ans := 0
	for i, item := range uf.root {
		if i == item {
			ans += 1
		}
	}
	return ans
}

// Print
func (uf *InnoUnionFindByRank) PrintRoots() {
	fmt.Println(uf.root)
}
